Sources/SwiftProtobuf/Message+StreamDecodingIterator.swift

```diff
@@ -0,0 +1,131 @@
+// Sources/SwiftProtobuf/Message+StreamDecodingIterator.swift - Iterator over binary delimited input streams
+//
+// Copyright (c) 2023 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Extensions to `Message` to provide binary coding and decoding using ``Foundation/Data``.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+extension Message {
+  
+  public static func streamDecodingIterator(inputStream: InputStream, bufferLength: Int = 32768) -> StreamDecodingIterator<Self> {
+    return StreamDecodingIterator<Self>(inputStream: inputStream, bufferLength: bufferLength)
+  }
+}
+
+/**
+ Iterates over a binary-delimited protobuf input stream
+ */
+public struct StreamDecodingIterator<M: Message>: IteratorProtocol {
+  
+  fileprivate let buffer: ResizingBuffer
+  
+  public init(inputStream: InputStream, bufferLength: Int = 32768) {
+    buffer = ResizingBuffer(inputStream: inputStream, bufferLength: bufferLength)
+  }
+  
+  public func next() -> M? {
+    do {
+      let messageLength = try decodeVarint()
+      if (messageLength == 0) {
```

Comment:
So this means a message that has all default values can never be read back on the stream, all default values encodes to zero length.

You'll likely need to let decodeVarint return errors, no data, and/or and explicit zero to that message case can be handled.  Maybe just returning a optional would do it, so if it return nil you know something went wrong?

Reply:
Good catch, I think it makes sense to make ```decodeVarint``` return nil on a zero-length read. Attempting a read seems to be the only way to know we're at the end an ```InputStream``` and in normal execution that will happen when attempting to read the varint data.